File: cpp/base/src/main/vsdk/toolkit/environment/geometry/element/Ray.cpp
#include <cmath>
#include <cstring>

#include "java/lang/String.h"
#include "vsdk/toolkit/common/VSDK.h"
#include "vsdk/toolkit/environment/geometry/element/Ray.h"
const double Ray::UNIT_DIRECTION_TOLERANCE = 1e-12;

Ray::Ray() : Ray(Vector3Dd(0, 0, 0), Vector3Dd(1, 0, 0), 0.0) {}

Ray::Ray(const Vector3Dd& origin, const Vector3Dd& direction)
    : Ray(origin, direction, 0.0) {}

Ray::Ray(const Vector3Dd& origin, const Vector3Dd& direction, double t)
    : origin(origin), direction(normalizeDirection(direction)), t(t) {}

Ray::Ray(const Ray& b) : origin(b.origin), direction(b.direction), t(b.t) {}

Ray Ray::copyOf(const Ray& other)
{
    return other;
}

Ray Ray::withOrigin(const Vector3Dd& newOrigin) const
{
    return Ray(newOrigin, direction, t);
}

Ray Ray::withDirection(const Vector3Dd& newDirection) const
{
    return Ray(origin, newDirection, t);
}

Ray Ray::withT(double newT) const
{
    if ( newT == t ) {
        return *this;
    }
    return Ray(origin, direction, newT);
}

void Ray::setOrigin(const Vector3Dd& newOrigin)
{
    origin = newOrigin;
}

void Ray::setDirection(const Vector3Dd& newDirection)
{
    direction = newDirection;
}

void Ray::setT(double newT)
{
    t = newT;
}

void Ray::setOriginAndDirection(
    const Vector3Dd& newOrigin, const Vector3Dd& newDirection)
{
    origin = newOrigin;
    direction = newDirection;
}

Vector3Dd Ray::normalizeDirection(const Vector3Dd& direction)
{
    double lengthSquared = direction.dotProduct(direction);
    if ( lengthSquared <= VSDK::EPSILON ) {
        return direction;
    }
    if ( std::abs(lengthSquared - 1.0) <= UNIT_DIRECTION_TOLERANCE ) {
        return direction;
    }
    return direction.multiply(1.0 / std::sqrt(lengthSquared));
}

bool Ray::equals(const Ray& other) const
{
    return t == other.t && origin == other.origin && direction == other.direction;
}

static unsigned int hashDouble(double val) {
    unsigned char bytes[sizeof(double)];
    memcpy(bytes, &val, sizeof(double));
    unsigned int h = 0u;
    for (int i = 0; i < (int)sizeof(double); ++i)
        h = h * 31u + static_cast<unsigned int>(bytes[i]);
    return h;
}

int Ray::hashCode() const
{
    unsigned int result = hashDouble(t);
    result = 31u * result + static_cast<unsigned int>(origin.hashCode());
    result = 31u * result + static_cast<unsigned int>(direction.hashCode());
    return static_cast<int>(result);
}

/**
Provides an object to text report conversion, optimized for human
readability and debugging. Do not use for serialization or persistence
purposes.
@return human-readable representation of current Ray
*/
java::String Ray::toString() const
{
    return "Ray Origin: <" + VSDK::formatDouble(origin.x()) + ", " +
        VSDK::formatDouble(origin.y()) + ", " +
        VSDK::formatDouble(origin.z()) + ">; Direction: <" +
        VSDK::formatDouble(direction.x()) + ", " +
        VSDK::formatDouble(direction.y()) + ", " +
        VSDK::formatDouble(direction.z()) + "> T: " +
        VSDK::formatDouble(t);
}

File: typescript/base/src/main/vsdk/toolkit/environment/geometry/element/Ray.ts
import { FundamentalEntity } from "../../../common/FundamentalEntity.js";
import { Vector3Dd } from "../../../common/linealAlgebra/Vector3Dd.js";
import { VSDK } from "../../../common/VSDK.js";
import type { GeometryRay } from "../Geometry.js";
import { Double } from "../../../../../java/lang/Double.js";
export class Ray extends FundamentalEntity implements GeometryRay {
    private static readonly UNIT_DIRECTION_TOLERANCE = 1e-12;
    private readonly origin: Vector3Dd;
    private readonly direction: Vector3Dd;
    private readonly t: number;
    public constructor();
    public constructor(origin: Vector3Dd, direction: Vector3Dd, t?: number);
    public constructor(ray: Ray);
    public constructor(a: Vector3Dd | Ray = new Vector3Dd(), b = new Vector3Dd(1, 0, 0), t = 0) {
        super();
        if (a instanceof Ray) {
            this.origin = a.origin;
            this.direction = a.direction;
            this.t = a.t;
        } else {
            this.origin = a;
            this.direction = Ray.normalizeDirection(b);
            this.t = t;
        }
    }
    public static copyOf(x: Ray) {
        if (x === null) throw new TypeError("Ray to copy cannot be null");
        return x;
    }
    public withOrigin(x: Vector3Dd) {
        return new Ray(x, this.direction, this.t);
    }
    public withDirection(x: Vector3Dd) {
        return new Ray(this.origin, x, this.t);
    }
    public withT(x: number) {
        return Double.compare(x, this.t) === 0 ? this : new Ray(this.origin, this.direction, x);
    }
    public getOrigin() {
        return this.origin;
    }
    public getDirection() {
        return this.direction;
    }
    public getT() {
        return this.t;
    }
    public equals(x: unknown) {
        return (
            x instanceof Ray &&
            Double.compare(this.t, x.t) === 0 &&
            this.origin.equals(x.origin) &&
            this.direction.equals(x.direction)
        );
    }
    public hashCode(): number {
        let result = 1;
        result = (Math.imul(31, result) + this.origin.hashCode()) | 0;
        result = (Math.imul(31, result) + this.direction.hashCode()) | 0;
        return (Math.imul(31, result) + Double.hashCode(this.t)) | 0;
    }
    private static normalizeDirection(direction: Vector3Dd): Vector3Dd {
        const lengthSquared = direction.dotProduct(direction);
        if (lengthSquared <= VSDK.EPSILON || Math.abs(lengthSquared - 1) <= this.UNIT_DIRECTION_TOLERANCE)
            return direction;
        return direction.multiply(1 / Math.sqrt(lengthSquared));
    }
    public override toString() {
        return `Ray Origin: ${this.origin}; Direction: ${this.direction} T: ${VSDK.formatDouble(this.t)}`;
    }
}
